llvm/test/CodeGen/RISCV/rv64zba.ll

```diff
@@ -2891,8 +2890,7 @@ define i64 @srli_slliw(i64 %1) {
 ;
 ; RV64ZBA-LABEL: srli_slliw:
 ; RV64ZBA:       # %bb.0: # %entry
-; RV64ZBA-NEXT:    slli a0, a0, 2
-; RV64ZBA-NEXT:    srli a0, a0, 4
+; RV64ZBA-NEXT:    srli a0, a0, 2
 ; RV64ZBA-NEXT:    slli.uw a0, a0, 4
 ; RV64ZBA-NEXT:    ret
 entry:
```

Comment:
Isn't this IR non-canonical per InstCombine

Reply:
The original test case is canonical: https://godbolt.org/z/ee3YPfY4s
```
define ptr @test(ptr %0, i64 %1) {
entry:
  %2 = lshr exact i64 %1, 2
  %3 = and i64 %2, 4294967295
  %4 = getelementptr inbounds i8, ptr %0, i64 600
  %5 = getelementptr [80 x i8], ptr %4, i64 %3
  ret ptr %5
}
```

Reply:
Yes with a GEP its canonical, but with a shl it isn't.

Reply:
Yeah, this was excessive reduction apparently.  

The fact instcombine prefers the opposite form does hint that we should maybe (also?) do the late match.  I was really hoping not to have to write that code...

Reply:
I just posted it. We had the majority of the code already. Is there a 3 shift version of this we should do without Zba?